src/libraries/Native/Unix/System.Native/pal_io.c

```diff
@@ -920,8 +920,17 @@ int32_t SystemNative_Poll(PollEvent* pollEvents, uint32_t eventCount, int32_t mi
     }
 
 
-    int useStackBuffer = bufferSize <= 2048;
-    struct pollfd* pollfds = (struct pollfd*)(useStackBuffer ? alloca(bufferSize) : malloc(bufferSize));
+    char stackBuffer[2048];
```

Comment:
Why not allocate a `struct pollfd` directly instead of doing these casts?  Might even defer the multiply-with-overflow-check to a call to `calloc()`.  If `SystemNative_Poll()` is mostly moved to an auxiliary function, the stack doesn't even need to be moved to allocate the 2048 bytes every time if the chunk of memory is allocated in the heap:

```C
int32_t SystemNative_Poll(PollEvent* pollEvents, uint32_t eventCount, int32_t milliseconds, uint32_t* triggered)
{
    static const uint32_t mallocThreshold = (uint32_t)(2048 / sizeof(struct pollfd));

    if (eventCount > mallocThreshold)
    {
                struct pollfd *pollfds = calloc(eventCount, sizeof(*pollfds));

                if (!pollfds) return Error_ENOMEM;

                int32_t ret = PollInternal(pollEvents, pollfds, eventCount, milliseconds, triggered);
                free(pollfds);
                return ret;
    }

    struct pollfds pollfds[mallocThreshold];
    return PollInternal(pollEvents, pollfds, eventCount, milliseconds, triggered);
}
```

Reply:
Thanks! However, this leads to:
 
    /home/omajid/devel/dotnet/runtime/src/libraries/Native/Unix/System.Native/pal_io.c(931,27): error GEF718E4C: variable length array folded to constant array as an extension [-Werror,-Wgnu-folding-constant] [/home/omajid/devel/dotnet/runtime/src/libraries/Native/build-native.proj]                                                                                                                                                                                                     
      struct pollfd pollfds[mallocThreshold];           
                            ^~~~~~~~~~~~~~~ 

Should I disable this warning?

Reply:
Ah, that's pseudo-code.  You can probably use a `#define` here to work around the warning instead of disabling the warning.

Reply:
> the stack doesn't even need to be moved to allocate the 2048 bytes every time if the chunk of memory is allocated in the heap

It is not how to C compilers work. It does not matter that you have the variable in the middle of the method. The stack for it will get allocate in the method prolog regardless.

> This also removes some of the branches in the actual function doing the work

It replaces the branching with a call that is likely going to be more expensive.


If you have split the function into two to make the code style better (personally, I liked the single method better), then it is ok. But if you have split the function into two to make it more efficient, I believe that it is doing exact opposite.

Reply:
It sounds like I should revert my patch to be the original/more-minimal version? And any correctness, style or performance issues that also affected the original code should be handled separately?

Reply:
Replacing malloc with calloc is good, but I would undo the split into two methods.